platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientProvidedPhasedActionRunner.java

```diff
@@ -74,8 +80,9 @@ public Object getBuildFinishedAction() {
 
         @Override
         public void collectActionResult(SerializedPayload serializedResult, PhasedActionResult.Phase phase) {
-            PhasedBuildActionResult res = new PhasedBuildActionResult(serializedResult, phase);
-            buildEventConsumer.dispatch(res);
+            PhasedBuildActionResult phaseResult = new PhasedBuildActionResult(serializedResult, phase);
+            BuildEventConsumer buildEventConsumer = ClientProvidedPhasedActionRunner.this.buildEventConsumer;
+            sideEffectExecutor.runIsolatableSideEffect(() -> buildEventConsumer.dispatch(phaseResult));
```

Comment:
ℹ️ As a nice side effect (pun intended) of the more general approach, dispatching the results of each phase of the phased action is now also a model side effect.